### PR TITLE
Clean Openacc code

### PR DESCRIPTION
**Pull request by Congyuwang**
It seems to me open-acc code contains quite a bit of confusing commands.

- `buffer` clone: seems to me no need to do this memory copy.
- `update directive`: seems to me no need to update, as after the data being just copied to device host there is no new update.
- `copyout`: I doubt `copyout` is correctly used: please refer to [openacc-guide.pdf](https://www.openacc.org/sites/default/files/inline-files/openacc-guide.pdf): "copyout - Create space for the listed variables on the device but do not initialize them. At the end of the region, copy the results back to the host and release the space on the device."
- `parallel` region: we have only a single parallel loop, so just use `acc parallel loop`.
- `present`: unnecessary following `copyin`.

### CODE REVIEW

File: project1/src/gpu/openacc_PartA.cpp [hunk 1]
```diff
@@ -28,39 +28,30 @@ int main(int argc, char **argv)
...
         for (int i = 0; i < width * height; i++)
         {
             grayImage[i] = (0.299 * buffer[i * num_channels] +
                             0.587 * buffer[i * num_channels + 1] +
                             0.114 * buffer[i * num_channels + 2]);
         }
-    }
-    auto end_time = std::chrono::high_resolution_clock::now();
-#pragma acc update self(grayImage[0 : width * height], \
-                        buffer[0 : width * height * num_channels])
```
Comment (Congyuwang) [inline project1/src/gpu/openacc_PartA.cpp:58]: unnecessary update

File: project1/src/gpu/openacc_PartA.cpp [hunk 2]
```diff
@@ -28,39 +28,30 @@ int main(int argc, char **argv)
...
-    for (int i = 0; i < width * height * num_channels; i++)
-    {
-        buffer[i] = input_jpeg.buffer[i];
-    }
-#pragma acc enter data copyin(grayImage[0 : width * height], \
-                              buffer[0 : width * height * num_channels])
+    std::chrono::milliseconds elapsed_time;
 
-#pragma acc update device(grayImage[0 : width * height], \
-                          buffer[0 : width * height * num_channels])
```
Comment (Congyuwang) [inline project1/src/gpu/openacc_PartA.cpp:41]: unnecessary update

File: project1/src/gpu/openacc_PartA.cpp [hunk 3]
```diff
@@ -28,39 +28,30 @@ int main(int argc, char **argv)
...
             grayImage[i] = (0.299 * buffer[i * num_channels] +
                             0.587 * buffer[i * num_channels + 1] +
                             0.114 * buffer[i * num_channels + 2]);
         }
-    }
-    auto end_time = std::chrono::high_resolution_clock::now();
-#pragma acc update self(grayImage[0 : width * height], \
-                        buffer[0 : width * height * num_channels])
-
-#pragma acc exit data copyout(grayImage[0 : width * height])
```
Comment (Congyuwang) [inline project1/src/gpu/openacc_PartA.cpp:60]: copyout not used as such.

File: project1/src/gpu/openacc_PartA.cpp [hunk 4]
```diff
@@ -28,39 +28,30 @@ int main(int argc, char **argv)
     int width = input_jpeg.width;
     int height = input_jpeg.height;
     int num_channels = input_jpeg.num_channels;
+    unsigned char *buffer = input_jpeg.buffer;
     unsigned char *grayImage = new unsigned char[width * height];
-    unsigned char *buffer = new unsigned char[width * height * num_channels];
-    for (int i = 0; i < width * height * num_channels; i++)
-    {
-        buffer[i] = input_jpeg.buffer[i];
-    }
```
Comment (Congyuwang) [inline project1/src/gpu/openacc_PartA.cpp:36]: unnecessary buffer clone?

File: project1/src/gpu/openacc_PartA.cpp [hunk 5]
```diff
@@ -28,39 +28,30 @@ int main(int argc, char **argv)
...
+// create space and copy to device without write to host
+#pragma acc data copyin(buffer[0 : width * height * num_channels])
 
-    auto start_time = std::chrono::high_resolution_clock::now();
-#pragma acc parallel present(grayImage[0 : width * height],             \
-                             buffer[0 : width * height * num_channels]) \
-    num_gangs(1024)
+// create space and copy to device, and write to host at the end of data scope
+#pragma acc data copy(grayImage[0 : width * height])
     {
```
Comment (Congyuwang) [inline project1/src/gpu/openacc_PartA.cpp:40]: create a data region, where grayImage get copied back to host at the end of which.

**Comment by tonyyxliu** [comment]
Thanks for pointing out!

@120090162 
Yebin, could you please take a look at this PR and see if it is possible to merge to our project?

**Comment by 120090162** [comment]
The code change made by you may have some bugs, the issue is like this:
srun: error: node01: task 0: Segmentation fault (core dumped)

the purpose of the PartA is to make students learn the calculated process of openACC, so I make the steps clearer than using one pragma to confuse students.

**Comment by Congyuwang** [comment]
Works flawlessly in my case.

Good that you have these pedagogical considerations.

I don't see how adding many of these unnecessary steps helps students understand how to code openACC better.

File: project1/src/gpu/openacc_PartA.cpp (hunk 1 above)
Comment (Congyuwang) [reply project1/src/gpu/openacc_PartA.cpp:58]: How is this resolved @120090162 ?

File: project1/src/gpu/openacc_PartA.cpp (hunk 4 above)
Comment (Congyuwang) [reply project1/src/gpu/openacc_PartA.cpp:36]: Add this?

**Comment by Congyuwang** [comment]
If this PR is not accepted, I do at least hope @120090162 improve upon the code. It seems loosely written. For the benefit of future students.

**Comment by yigudove** [comment]
I think some points here are not that unnecessary as it is tutorial code. It's sometimes strange to criticize a tutorial code is not 'clean'.

**Comment by 120090162** [comment]
It seems that there was something wrong with the cluster just now, your code is correct.
But, you should note that the sample code is just for teaching. The sample code needs to clearly demonstrate the process of openACC acceleration rather than using simplified versions of commands.
Overall, thank you for your contributions.

**Comment by Congyuwang** [comment]
> I think some points here are not that unnecessary as it is tutorial code. It's sometimes strange to criticize a tutorial code is not 'clean'.

I partially agree. For example, the 'present' clause might actually be a good practice since it double checks to ensure the memory is actually allocated on the device (although 'copyxx' already does the same checks). I opened this PR with the intention of discussion and modification. Now that it is accepted anyway.

**Pull request merged**